Automattic/wp-calypso
# Analytics: Registration, Signup, AddToCart, Purchase events updates

### Updates included in this Pull Request

* Pinterest new pixels:
  * Registration: uses `lead` event
  * Signup: uses `signup` event
  * AddToCart: uses `addtocart` event
  * Purchase: uses `checkout` event
  * Pageview/Retarget
* Removed legacy trackers (previously already disabled): 
  * Yandex
  * Yahoo legacy
  * Media Wallah
* Google Ads Gtag:
  * Registration: `AW-946162814/_6cKCK6miZYBEP6YlcMD`
  * AddToCart: [WPCom] `AW-946162814/MF4yCNi_kZYBEP6YlcMD`
  * AddToCart: [Jetpack] `AW-937115306/BrwBCIWziZYBEKr97L4D`
  * Fixed retargeting by implementing virtual page update for both WPCom and Jetpack pixels
* Bing pixels added:
  * Registration
  * AddToCart
  * Retargeting/pageview fixed*
* Facebook pixels:
  * Registration: uses `Lead` FB event.
* Floodlight new pixels:
  * Registration
  * AddToCart
  * Updated `http://` to `https://wordpress.com` in every pixel we have in the Floodlight manger
  * Fixed Alerts in Floodlight configuration due to missing `num` parameters
  * Checked and fixed every single event we have to explicitly set `ord` and `num` according to its event/counter type
* Quantcast new pixel:
  * Signup
* Twitter pixels:
  * Signup new pixel
  * Purchase reconfigured to fire once per purchase
* Google Analytics:
  * Fixed 0-value Purchase events
* Fixed non normalized emails in various trackers
* Fixes the way events are handled while trackers are loaded to enqueue them instead of discarding all events but the first one.
* Fixes potential use of uninitialized trackers (by using the queue).
* Ensures the GDPR cookies and `isAdTrackingAllowed()` are properly checked before all events.
* Debug: moved `debug()` before firing event to provide more meaningful context in case of error
* Debug: print full tracker list of parameters by using `...params` notation where it makes sense
* Debug: dump `dataLayer` also on `retarget()` so we can inspect Gtag proper operation

### Testing WPCom

**Testing Setup**

- Make sure you have the following features enabled in:
  - `development.json`: `gdpr-banner`, `google-analytics`, `ad-tracking` 
  - or via the URL params in the calypso.live URL: https://calypso.live/?branch=add/update-ad-trackers (`?flags=gdpr-banner,google-analytics,ad-tracking`)

- Enable debug logs by entering the following in your browser console:
```
localStorage.setItem( 'debug', 'calypso:analytics:*' );
```
- Set the following to ensure the GDPR logic does not block the trackers:
```
document.cookie = 'sensitive_pixel_option=yes;'
```

- Reload Calypso and make sure you see `isAdTrackingAllowed: true` through out the test in the JS console.

- Make sure to select the options "Preserve Logs" in both Chrome's JS console and Network tab.

- While testing, especially when filtering the JS console make sure to look back at the unfiltered log every now and then to check if there's any suspicious JavaScript error popping up.

**Page view**

- Making sure ad tracking is enabled, clean your JS history and reload ("Empty Cache and Hard Reload") the page. Ghostery should report around 25 trackers.

![image](https://user-images.githubusercontent.com/10284338/53736992-2f7efc80-3e83-11e9-9f32-e1ab3be4fcf5.png)

- Filtering by `retarget` will give you a list of each pixel that's been fired. To double check, look in the Network tab for a corresponding activity. All should have a corresponding activity except the Gtag one which get written in the `dataLayer`.

- Note: some of the trackers are marked with `[rate limited]`, which means that they will not fire in the next 24 hours when clicking on a new page, but they will if you reload Calypso.

**Registration**

- Open a new incognito window, make sure to enable tracking and logging as explained in the beginning (you should see `isAdTrackingAllowed: true` when reloading).

- Create a new user and a new test website with a free plan: `/start/user?flags=gdpr-banner,google-analytics,ad-tracking`.

- After submitting your user and email information filter by `recordRegistration` and you should see something like:

![image](https://user-images.githubusercontent.com/10284338/53738138-4bd06880-3e86-11e9-85c8-26b7f33433eb.png)

- Double check those events have a corresponding entry in the Network tab and the `send_to` parameter in the `dataLayer` for the last event corresponds to the Registration pixel in the code.

**Signup**

- After completing the signup process (select free plan) filter by `recordSignup` and you should see something like this:

![image](https://user-images.githubusercontent.com/10284338/53805031-fe650180-3f40-11e9-8e08-6b4ef03560e5.png)

- Double check those events have a corresponding entry in the Network tab and the last `send_to` parameter in the `dataLayer` for the last event corresponds to the Signup pixel in the code.

**AddToCart**

- Log in to your newly created account and select a paid plan. Filter the JS console with `recordAddToCart`, you should see something like:

![image](https://user-images.githubusercontent.com/10284338/53805821-1fc6ed00-3f43-11e9-81d9-f6be3bac15c5.png)

- The `dataLayer` dump should show a `"send_to": "AW-946162814/MF4yCNi_kZYBEP6YlcMD"` which should correspond to `wpcomGoogleAdsGtagAddToCart` in `ad-tracking.js`.

- Double check those events have a corresponding entry in the Network tab.

**Purchase**

- After purchasing a paid plan you should see the following trackers when filtering by `recordOrder`:

![image](https://user-images.githubusercontent.com/10284338/53827699-9bd92900-3f73-11e9-9998-bce2b78443c7.png)

- If you purchased a WPCom plan you should see something similar to the following entry in the `dataLayer`. The `send_to` param corresponds to the `wpcomGoogleAdsGtagPurchase` in the code.

![image](https://user-images.githubusercontent.com/10284338/53827764-c75c1380-3f73-11e9-8b1d-7b549b517eef.png)


### Testing Jetpack

- Purchase a Jetpack plan, make sure there are no errors in the JS console, filter by `recordOrder`, you should see something like this:

- Make sure those pixels actually exist in the Network tab without errors:

![image](https://user-images.githubusercontent.com/10284338/53829707-a77b1e80-3f78-11e9-9552-81f685d9952d.png)

![image](https://user-images.githubusercontent.com/10284338/53829749-bd88df00-3f78-11e9-8ba2-c7f48c89b579.png)

- Jetpack Gtag `jetpackGoogleAdsGtagPurchase`:

![image](https://user-images.githubusercontent.com/10284338/53829768-ca0d3780-3f78-11e9-9e06-35fada07dc87.png)
